openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMBinaryClassificationModelTrainer.java

```diff
@@ -343,42 +400,35 @@ static void saveModelFileToDisk(final SWIGTYPE_p_void swigBoosterHandle, final P
      *
-     * @param dataset              Input train dataset (with target label)
-     * @param swigFeatureDataArray SWIG array of doubles to which the features data will be copied into (in row-major format).
-     * @param swigLabelDataArray   SWIG array of floats to which the labels shall be copied.
+     * @param dataset       Input train dataset (with target label)
+     * @param swigTrainData SWIGTrainData object.
      */
     private static void copyTrainDataToSWIGArrays(final Dataset dataset,
-                                                  final SWIGTYPE_p_double swigFeatureDataArray,
-                                                  final SWIGTYPE_p_float swigLabelDataArray) {
+                                                  final SWIGTrainData swigTrainData) {
 
         final DatasetSchema datasetSchema = dataset.getSchema();
         final int numFields = datasetSchema.getFieldSchemas().size();
-        final int numFeatures = datasetSchema.getPredictiveFields().size();
-        // if target index doesn't exit, return -1.
         final int targetIndex = datasetSchema.getTargetIndex().orElse(-1);
 
         final Iterator<Instance> iterator = dataset.getInstances();
-        long rowIdx = 0;
         while (iterator.hasNext()) {
             final Instance instance = iterator.next();
 
-            // Set the label value for this instance:
-            lightgbmlib.floatArray_setitem(swigLabelDataArray, rowIdx, (float) instance.getValue(targetIndex));
-
-            // Set the features values for this instance:
-            final long rowOffset = rowIdx * numFeatures;
-            for (int colIdx = 0, afterTargetColOffset = 0; colIdx < numFields; ++colIdx) {
-                if (colIdx == targetIndex) {
-                    afterTargetColOffset = -1; // Initially 0, becomes -1 after passing the target column.
-                } else {
-                    lightgbmlib.doubleArray_setitem(
-                            swigFeatureDataArray,
-                            rowOffset + colIdx + afterTargetColOffset,
-                            instance.getValue(colIdx)
-                    );
+            swigTrainData.addLabelValue((float) instance.getValue(targetIndex));
+
+            for (int colIdx = 0; colIdx < numFields; ++colIdx) {
+                if (colIdx != targetIndex) {
+                    swigTrainData.addFeatureValue(instance.getValue(colIdx));
                 }
             }
-            ++rowIdx;
+        }
+        logger.debug("Copied train data of size {} into {} chunks.",
+                swigTrainData.swigLabelsChunkedArray.get_add_count(),
+                swigTrainData.swigLabelsChunkedArray.get_chunks_count()
+        );
+        if (swigTrainData.swigLabelsChunkedArray.get_add_count() == 0) {
+            logger.error("Received empty train dataset!");
+            throw new RuntimeException("Received empty train dataset for LightGBM!");
```

Comment:
Should you throw a `ModelTrainingException` and allow some error handling upstream?

Reply:
What kind of error handling should we do?

Reply:
@JPDSousa any inputs here?

Reply:
It's not clear to me how we can recover from this error. We'll need to wait for João to clarify

Reply:
I'm not considering necessarily a recovery flow but instead treating a well-identified error differently from other Runtime Exceptions. Since at the moment we are not doing it, and to avoid throwing a checked exception, at least we can try to be a little more specific and throw a subclass of `RuntimeException` (e.g.: `IllegalArgumentException` other examples at https://docs.oracle.com/javase/7/docs/api/java/lang/RuntimeException.html)

Reply:
IllegalArgumentException seems the right fit ;)